{% extends 'base.html' %}

{% block head %}
<style>
    hr {
        width: 400px;
        margin-left: 0;
    }

    /* Via https://codepen.io/SaraSoueidan/pen/CyFbd */
    .quotes {
        position: absolute;
        color: #132b44;
        font-size: 3em;
    }

    .leftquote {
        top: .5rem;
        left: -.25rem;
    }

    .rightquote {
        bottom: 1rem;
        right: -.25rem;
    }
</style>
{% endblock %}

{% block content %}
<div style="padding-top:30px; padding-bottom:30px; padding-left:30px">
    <h2>About DMS MaP experiments</h2>
    <hr>
    <div style="padding-left:50px; padding-right:20px; padding-top:18px; line-height:18px;">
        <p>
            DMS has emerged as one of the pre-eminent choices for RNA structure determination.
            DMS can be added to cells, tissues, or in vitro solution, and it rapidly and specifically
            modifies solvent accessible adenines and cytosines at their Watsonâ€“Crick base-pairing
            positions. In standard experimental conditions, an accessible nucleotide has ~2%
            chance of reacting with DMS, which results in multiple DMS modifications per single RNA
            molecule. The DMS modifications therefore report on the folding of each individual RNA
            molecule.
        </p>
        <div style="text-align: center">
            <img style="margin-bottom: 1rem; width: 100%; max-width: 500px;" src="/static/dms_image.png">
        </div>
        <p>
            DMS mutational profiling with sequencing (DMS-MaPseq), encodes DMS modifications as mismatches that get
            incorporated during reverse transcription by a thermostable group II intron reverse transcriptase (TGIRT).
            Due to the high fidelity of TGITR, the background incorporation of mismatches is typically lower than
            sequencing error. Thus, the observed rate of a mismatch at a given nucleotide is directly
            proportional to its DMS reactivity. A big advantage of DMS-MaPseq is that any RNA of interest can be
            targeted
            for library generation and analyses using sequence specific primers.
        </p>
        <div style="text-align: center">
            <img style="margin-bottom: 1rem; width: 100%; max-width: 500px;" src="/static/tigrt_image.png">
        </div>
    </div>
    <h2>About the DREEM analysis</h2>
    <div style="padding-left:50px; padding-right:20px; padding-top:18px; line-height:18px;">
        <hr>
        <p>
            The DREEM web tool provides a simple platform for analyzing DMS-reactivity of an RNA. The user input is a
            raw sequencing file (.fastq) generated from a DMS-MaPseq experiment,  and a sequence of the RNA of interest
            (.fasta). The DREEM algorithm performs sequence alignment using bowtie-2 and outputs the mismatch rate per
            nucleotide.
        </p>
    </div>
</div>


{% endblock content %}

{% block bodyend %}
{% endblock %}