<link rel="stylesheet" href="../static/css/theme.css">{% extends 'base.html' %}

{% block head %}
<style>
    input {
        font-family: 'Cousine', monospace;
    }

    .fileUpload {
        position: relative;
        overflow: hidden;
    }

    .fileUpload input.upload {
        position: absolute;
        top: 0;
        right: 0;
        margin: 0;
        padding: 0;
        font-size: 20px;
        cursor: pointer;
        opacity: 0;
        filter: alpha(opacity=0);
    }
</style>
{% endblock head %}

{% block content %}
<div class="jumbotron">
    <h1>Welcome to the DREEM webserver</h1>
    <p>
        DREEM is a set of computational tools to analyze chemical probing data from dimethyl sulfate mutational profiling with
        sequencing (DMS-MaPseq). The tool takes as input raw sequencing files and target RNA sequence and outputs the
        DMS reactivity per nucleotide, which is used to determine the RNA structure. This website is free and open to all
        users and there is no login requirement.
    </p>
    <div style="text-align: center; margin-top: 25px;">
        <img class="d-none d-md-inline" src="/static/dreem_front_page.png" width="100%;">
        <img class="d-inline d-md-none" src="/static/dreem_front_page.png" width="100%;">
    </div>
</div>
<div class="bd-example">
    <h2>Submit your mutational analysis profiling (MaP) job</h2>
    <p>
        This server generates the population average for MaP experiments.
        The server accepts a FASTA-formatted file containing the sequences of target RNAs as well as FASTQ-formatted files containing
        the raw sequencing reads from an illumina-based sequencer.
        To limit how long it takes to run jobs, this webserver has a limit of 1000 reference sequences. We recommend ziping your fastq
        files if they are larger 50 Mb. If dealing with very large fastq files please consider <a href="https://github.com/jyesselm/dreem">
        using a local copy of DREEM</a>.
        Don't know where to start? <a href="/download/example-input">Download example input</a>
    </p>
    <form id="submission" class="form-horizontal" method='post' action='/request' enctype="multipart/form-data">
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fasta file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fasta_file_field" class="form-control" placeholder="Choose File" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fasta_file" type="file" class="form-control-file upload" required/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fasta formated file that contains all reference sequences. <a href="#fasta_format">see examples</a>
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fastq1 file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fastq1_file_field" class="form-control" placeholder="Choose File" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fastq1_file" type="file" class="form-control-file upload" required/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fastq formatted file containing sequence reads in the forward direction. Can be zip compressed. <a href="#fastq_format">see example</a>
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fastq2 file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fastq2_file_field" class="form-control" placeholder="Choose File (Optional)" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fastq2_file" type="file" class="form-control-file upload"/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fastq formatted file containing sequence reads in the reverse direction. Can be zip compressed.
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Job Name</label>
            <div class="col-sm-8">
                <input type="text" class="form-control" name="name" placeholder="Optional">
                <div class="invalid-feedback" for="name"></div>
                <small class="form-text text-muted">
                    Provide a name for your submission to help you more easily recognize it later
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Email</label>
            <div class="col-sm-8">
                <input type="email" class="form-control" name="email" placeholder="Optional">
                <div class="invalid-feedback" for="email"></div>
                <small class="form-text text-muted">
                    Provide your email to get a notification once the job is complete
                </small>
            </div>
        </div>
        <button type="submit" class="btn btn-primary">Submit</button>
        <a class="btn btn-secondary" href="/result/demo" role="button">Example Results</a>

    </form>
    <hr>
    <h2>Frequently asked questions (FAQ)</h2>
    <p><h5>Answers for the most common questions</h5></p>
    <ul>
        <li><a href="/about"><h5>What are MaP experiments?</h5></a></li>
        <li><a href="/about#analysis"><h5>How is the analysis performed?</h5></a></li>
        <li><a href="/about#ss_prediction"><h5>How to predict a secondary structure with MaP results?</h5></a></li>
        <li><a href="#fasta_format"><h5>What is the expected FASTA file format?</h5></a></li>
        <li><a href="#fastq_format"><h5>What is the expected FASTQ file format?</h5></a></li>
        <li><a href="#normalization"><h5>How to normalize MaP results?</h5></a></li>
    </ul>
    <br>
    <h3 id="too-many-reads">Upload is slow for large FASTQ files</h3>
    <p>This webserver can accept zipped FASTQ files. Consider zip compressing your files before submitting them to reduce upload time.
    </p>
    <h3 id="fasta_format">What is the expected FASTA file format?</h3>
    <p>FASTA format is a text-based format for representing DNA sequences. must formatted as below. With an '>' followed
        by name. <b>There should be no spaces in-between '>' and the sequence name</b>. Next line contains the DNA sequence (all Us should
        be Ts). See entire example below. <a href="/download/example-fasta-file">Download example FASTA file</a>.</p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        >p4-p6-m2seq<br>
        GGCCAAAACAACGGAATTGCGGGAAAGGGGTCAACAGCCGTTCAGTACCAAGTCTCAGGGGAAACTTTGAGATGGCCTTGCAAAGGGTATGGTAATAAGCTGACGGACATGGTCCTAACCACGCAGCCAAGTCCTAAGTCAACAGATCTTCTGTTGATATGGATGCAGTTCAAAACCAAACCAAAGAAACAACAACAACAAC
    </div>
    <br>
    <p>If you have multiple reference sequences must not contain empty lines</p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        >seq_1<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGATTAGCAAGGACATGCAGAGCAAGGGGGAAACTTCACCTCTGCAACAGCCACCTAGTCCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_2<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGTTAACACCCGATGATGGAAGGTAGGAGCAACGTTGGCAGGGGAAACTTTGCCAACGGCCTACTGGACATCGGCAAGTTAACCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_3<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGTGGATAGTGACATGAATTCTCAGGGGAAACTTTGAGAATTCAACAGCACAAGAAGCCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_4<br>
        .<br>
        .<br>
        .<br>
    </div>
    <h3 id="fastq_format">What is the expected FASTQ file format?</h3>
    <p>FASTQ format is a text-based format for storing DNA sequences with corresponding quality scores. These quality
        scores define how likely a given nucleotide is to be correctly read via sequencing. Below is an example of a properly
        formatted FASTQ file. Each read contains 4 lines. (1) the name of the read, (2) the read sequence, (3) '+', and (4) the
        read quality by nucleotide. <a href="/download/example-fastq-file">Download example FASTQ file</a>
    </p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        @FS10000899:22:BPG61606-0731:1:1101:1200:1000 2:N:0:1<br>
        GGAAGATCGAGTAGATCAAAGGACGTATGGCGGGCGGCCGAGACCCCGAACTACGAGGAACAGAGGAAACTCTACCCCTCGCGGGGTCGTTTGACGCCGCCCGCCTAAGCGT
        CCAAAGAAACAACAACAACAACAGCCACCCAGGCAGATC<br>
        +<br>
        FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
        FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF<br>
    </div><br>
    <h3 id="normalization">How to normalize MaP results?</h3>
    <p>If performing RT with the TGIRT or Marathon the background is low enough that no normalization is required. If
        chemical probing is done with SSII/mn2+ then normalization may be required. We suggest running two seperate jobs. One
        with the chemical modified data and one with a no-modification (nomod). With the data from both runs you can normalize
        either by dividing or substracting out the nomod condition. For more information please consult the following publications:
        <a href="https://www.pnas.org/doi/10.1073/pnas.1807988115">(1)</a> and <a href="https://www.nature.com/articles/nmeth.3029">(2)</a>
    </p>
    </div>
{% endblock content %}

{% block bodyend %}
<script>

  var fasta_file = $('[name="fasta_file"]');
  var fasta_file_field = $('[name="fasta_file_field"]');
  var fastq1_file = $('[name="fastq1_file"]');
  var fastq2_file = $('[name="fastq2_file"]');

  fasta_file.on('input', function () {
    document.getElementById("fasta_file_field").value = fasta_file.val().substring(12);
    this.setCustomValidity("");
    if (fasta_file[0].files.length === 0) {
      return;
    }
    var filename = fasta_file[0].files[0].name;
    var format = /[ `!@#$%^&*(){}]+/;
    if(format.test(filename)) {
        this.setCustomValidity("filename must not contain symbols other than _ or -");
        this.reportValidity();
    }
  });


  fastq1_file.on('input', function () {
    document.getElementById("fastq1_file_field").value = fastq1_file.val().substring(12);
    this.setCustomValidity("");
    if (fastq1_file[0].files.length === 0) {
      return;
    }
    var filename = fastq1_file[0].files[0].name;
    var format = /[ `!@#$%^&*(){}]+/;
    if(format.test(filename)) {
        this.setCustomValidity("filename must not contain symbols other than _ or -");
        this.reportValidity();
    }
    //alert(fasta_file[0].files[0].size);
  });

  fastq2_file.on('input', function () {
    document.getElementById("fastq2_file_field").value = fastq2_file.val().substring(12);
    this.setCustomValidity("");
    if (fastq2_file[0].files.length === 0) {
      return;
    }
    var filename = fastq2_file[0].files[0].name;
    var format = /[ `!@#$%^&*(){}]+/;
    if(format.test(filename)) {
        this.setCustomValidity("filename must not contain symbols other than _ or -");
        this.reportValidity();
    }
    //alert(fasta_file[0].files[0].size);
  });



</script>
{% endblock bodyend %}
