<link rel="stylesheet" href="../static/css/theme.css">{% extends 'base.html' %}

{% block head %}
<style>
    input {
        font-family: 'Cousine', monospace;
    }

    .fileUpload {
        position: relative;
        overflow: hidden;
    }

    .fileUpload input.upload {
        position: absolute;
        top: 0;
        right: 0;
        margin: 0;
        padding: 0;
        font-size: 20px;
        cursor: pointer;
        opacity: 0;
        filter: alpha(opacity=0);
    }
</style>
{% endblock head %}

{% block content %}
<div class="jumbotron">
    <h1>Welcome to the DREEM webserver</h1>
    <p>
        DREEM is a computational tool to analyze chemical probing data from dimethyl sulfate mutational profiling with
        sequencing (DMS-MaPseq). The tool takes as input raw sequencing files and target RNA sequence and outputs the
        DMS reactivity per nucleotide, which is used to determine the RNA structure.
    </p>
    <div style="text-align: center; margin-top: 25px;">
        <img class="d-none d-md-inline" src="/static/dreem_front_page.png" width="100%;">
        <img class="d-inline d-md-none" src="/static/dreem_front_page.png" width="100%;">
    </div>
</div>
<div class="bd-example">
    <h2>Submit your DREEM job</h2>
    <p>
        This server accepts a FASTA-formatted file containing the sequences of target RNAs as well as FASTQ-formatted files containing
        the raw sequencing reads from an illumina-based sequencer.
        To limit how long it takes to run jobs, this webserver has a limit of 100 reference sequences with 100 Mb sequencing read
        files, please <a href="https://github.com/jyesselm/dreem">use a local copy of DREEM</a>.
        Don't know where to start? <a href="/download/example-input">Download example input</a>
    </p>
    <form id="submission" class="form-horizontal" method='post' action='/request' enctype="multipart/form-data">
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fasta file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fasta_file_field" class="form-control" placeholder="Choose File" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fasta_file" type="file" class="form-control-file upload" required/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fasta formated file that contains all reference sequences. <a href="#fasta_format">see examples</a>
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fastq1 file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fastq1_file_field" class="form-control" placeholder="Choose File" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fastq1_file" type="file" class="form-control-file upload"/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fastq formatted file containing sequence reads in the forward direction <a href="#fastq_format">see example</a>
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Fastq2 file</label>
            <div class="col-sm-8">
                <div class="input-group">
                    <input id="fastq2_file_field" class="form-control" placeholder="Choose File (Optional)" disabled="disabled"
                           style="background-color : white">
                    <div class="input-group-btn">
                        <div class="fileUpload btn btn-success">
                            <span>Upload</span>
                            <input name="fastq2_file" type="file" class="form-control-file upload"/>
                        </div>
                    </div>
                </div>
                <small class="form-text text-muted">
                    A fastq formatted file containing sequence reads in the reverse direction
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Job Name</label>
            <div class="col-sm-8">
                <input type="text" class="form-control" name="name" placeholder="Optional">
                <div class="invalid-feedback" for="name"></div>
                <small class="form-text text-muted">
                    Provide a name for your submission to help you more easily recognize it later
                </small>
            </div>
        </div>
        <div class="form-group row">
            <label class="col-sm-2 col-form-label-lg">Email</label>
            <div class="col-sm-8">
                <input type="email" class="form-control" name="email" placeholder="Optional">
                <div class="invalid-feedback" for="email"></div>
                <small class="form-text text-muted">
                    Provide your email to get a notification once the job is complete
                </small>
            </div>
        </div>
        <button type="submit" class="btn btn-primary">Submit</button>
        <a class="btn btn-secondary" href="/result/demo" role="button">Example Results</a>

    </form>
    <hr>
    <h2>Frequently asked questions (FAQ)</h2>
    <p>Answers for the most common questions</p>
    <h3 id="too-many-reads">The server says my FASTQ files are too large</h3>
    <p>To avoid long waits we have set a limit on 100 Mb per FASTQ file. If you wish to test your data on the server
        there are multiple ways to reduce the size of your file for testing purposes. The simplest is to use the 'head'
        command which is builtin on all mac and linux computers. Open the terminal on your computer and navigate to the
        directory with your FASTQ files.
    </p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
       head -n 400000 test_1.fastq > test_1_trimmed.fastq
    </div><br>
    <p>This will take the first hundred thousand reads and move them into a new file that you can upload</p>
    <h3 id="fasta_format">What is the expected FASTA file format?</h3>
    <p>FASTA format is a text-based format for representing DNA sequences. must formatted as below. With an '>' followed
        by name. <b>There should be no spaces in-between '>' and the sequence name</b>. Next line contains the DNA sequence (all Us should
        be Ts). See entire example below. <a href="/download/example-fasta-file">Download example FASTA file</a>.</p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        >p4-p6-m2seq<br>
        GGCCAAAACAACGGAATTGCGGGAAAGGGGTCAACAGCCGTTCAGTACCAAGTCTCAGGGGAAACTTTGAGATGGCCTTGCAAAGGGTATGGTAATAAGCTGACGGACATGGTCCTAACCACGCAGCCAAGTCCTAAGTCAACAGATCTTCTGTTGATATGGATGCAGTTCAAAACCAAACCAAAGAAACAACAACAACAAC
    </div>
    <br>
    <p>If you have multiple reference sequences must not contain empty lines</p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        >seq_1<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGATTAGCAAGGACATGCAGAGCAAGGGGGAAACTTCACCTCTGCAACAGCCACCTAGTCCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_2<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGTTAACACCCGATGATGGAAGGTAGGAGCAACGTTGGCAGGGGAAACTTTGCCAACGGCCTACTGGACATCGGCAAGTTAACCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_3<br>
        GGAAGATCGAGTAGATCAAAGTTGATATGGTGGATAGTGACATGAATTCTCAGGGGAAACTTTGAGAATTCAACAGCACAAGAAGCCTAAGTCAACAAAGAAACAACAACAACAAC<br>
        >seq_4<br>
        .<br>
        .<br>
        .<br>
    </div>
    <h3 id="fastq_format">What is the expected FASTQ file format?</h3>
    <p>FASTQ format is a text-based format for storing DNA sequences with corresponding quality scores. These quality
        scores define how likely a given nucleotide is to be correctly read via sequencing. Below is an example of a properly
        formatted FASTQ file. Each read contains 4 lines. (1) the name of the read, (2) the read sequence, (3) '+', and (4) the
        read quality by nucleotide. <a href="/download/example-fastq-file">Download example FASTQ file</a>
    </p>
    <div class="container bg-light text-wrap" style="width: 1100px;word-wrap: break-word;">
        @FS10000899:22:BPG61606-0731:1:1101:1200:1000 2:N:0:1<br>
        GGAAGATCGAGTAGATCAAAGGACGTATGGCGGGCGGCCGAGACCCCGAACTACGAGGAACAGAGGAAACTCTACCCCTCGCGGGGTCGTTTGACGCCGCCCGCCTAAGCGT
        CCAAAGAAACAACAACAACAACAGCCACCCAGGCAGATC<br>
        +<br>
        FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
        FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF<br>
    </div>


    </div>
{% endblock content %}

{% block bodyend %}
<script>


  var seq = $('[name="sequence"]');
  var struct = $('[name="structure"]');
  var fasta_file = $('[name="fasta_file"]');
  var fastq1_file = $('[name="fastq1_file"]');
  var fastq2_file = $('[name="fastq2_file"]');

  function checkLengthMatch(dotbracket, sequence) {
    // We don't want to bother warning the user if they haven't filled the other one out yet - the `required` attr handles that
    if (sequence == '' || dotbracket == '') return '';
    if (sequence.length !== dotbracket.length) return 'Sequence and structure lengths must match';
    return '';
  }

  function getPairs(dotBracket) {
    var pairs = [];
    var unpaired = [];
    for (var i = 0; i < dotBracket.length; i++) {
      switch (dotBracket[i]) {
        case '.':
          break;
        case '(':
          unpaired.push(i);
          break;
        case ')':
          pairs.push([unpaired.pop(), i]);
          break;
      }
    }
    return pairs;
  }

  function isValidPair(baseA, baseB) {
    var iupac = {
      'A': ['A'],
      'C': ['C'],
      'G': ['G'],
      'U': ['U'],
      'W': ['A', 'U'],
      'S': ['C', 'G'],
      'M': ['A', 'C'],
      'K': ['G', 'U'],
      'R': ['A', 'G'],
      'Y': ['C', 'U'],
      'B': ['C', 'G', 'U'],
      'D': ['A', 'G', 'U'],
      'H': ['A', 'C', 'U'],
      'V': ['A', 'C', 'G'],
      'N': ['A', 'C', 'G', 'U']
    }
    var compliments = {
      'A': 'U',
      'C': 'G',
      'G': 'Y',
      'U': 'R',
      'W': 'D',
      'S': 'B',
      'M': 'K',
      'K': 'M',
      'R': 'Y',
      'Y': 'R',
      'B': 'N',
      'D': 'N',
      'H': 'D',
      'V': 'B',
      'N': 'N'
    }

    // If all the bases of one symbol can match with one base of the other symbol
    if (iupac[baseA].some(base => iupac[compliments[baseB]].includes(base))) return true;
    else return false;
  }

  function checkStructureValid(dotbracket) {
    var pairTally = 0;
    var loopTally = 0;
    var inPairOpening = true;
    for (var char of dotbracket) {
      switch (char) {
        case '(':
          pairTally += 1;
          inPairOpening = true;
          loopTally = 0;
          break;
        case ')':
          pairTally -= 1;
          if (inPairOpening && loopTally < 3) return 'Hairpin loops must be at least 3 bases long';
          inPairOpening = false;
          loopTally = 0;
          break;
        case '.':
          loopTally += 1;
          break;
      }

      if (pairTally < 0) return 'A ")" must not occur before having a matching "("';
    }

    if (pairTally > 0) return 'All "("s must have a matching ")"';
    return '';
  }

  function checkValidPairs(dotBracket, sequence) {
    var normalizedSequence = sequence.replace(/T/g, 'U');
    // We can't validate this if the dot bracket is wrong - the validation for that field will handle that
    if (checkStructureValid(dotBracket)) return '';
    var pairs = getPairs(dotBracket);
    for (var pair of pairs) {
      var baseA = normalizedSequence[pair[0]];
      var baseB = normalizedSequence[pair[1]];
      if (!isValidPair(baseA, baseB)) return 'Invalid pair: ' + baseA + baseB;
    }
    return '';
  }

  function validateSeq() {
    var dotBracket = struct.val();
    var sequence = seq.val().toUpperCase() || 'N'.repeat(dotBracket.length);

    seq[0].setCustomValidity(
      checkLengthMatch(dotBracket, sequence) ||
      checkValidPairs(dotBracket, sequence)
    );
  }

  function validateStruct() {
    var dotBracket = struct.val();
    var sequence = seq.val().toUpperCase() || 'N'.repeat(dotBracket.length);

    struct[0].setCustomValidity(
      // NOTE: This is also validated on the backend. If the maximum changes, make sure to change it there too
      dotBracket.length > 200 ?
        'Please enter a a structure with 200 or less nucleotides' : '' ||
        checkLengthMatch(dotBracket, sequence) ||
        checkStructureValid(dotBracket)
    );
  }

  fasta_file.on('change', function () {
    document.getElementById("fasta_file_field").value = fasta_file.val().substring(12);
    if (fasta_file[0].files.length === 0) {
      return;
    }
    //var reader = new FileReader();

    //alert(fasta_file[0].files[0].size);
  });

  fastq1_file.on('change', function () {
    document.getElementById("fastq1_file_field").value = fastq1_file.val().substring(12);
    if (fasta_file[0].files.length === 0) {
      return;
    }
    //alert(fasta_file[0].files[0].size);
  });

  fastq2_file.on('change', function () {
    document.getElementById("fastq2_file_field").value = fastq2_file.val().substring(12);
    if (fasta_file[0].files.length === 0) {
      return;
    }
    //alert(fasta_file[0].files[0].size);
  });


  seq.on('input', function () {
    // Run both in case changes in this field wind up impacting validity of the other field
    // in cases where they rely on one another
    validateSeq();
    validateStruct();
  });

  struct.on('input', function () {
    // Run both in case changes in this field wind up impacting validity of the other field
    // in cases where they rely on one another
    validateStruct();
    validateSeq();
  });

  // Custom error messages for pattern mismatch
  seq.on('invalid', function () {
    if (this.validity.patternMismatch) {
      this.setCustomValidity('Sequence must only contain IUPAC nucleic acid notation symbols (A, C, G, T, U, W, S, M, K, R, Y, B, D, H, V, and N)');
    }
  });

  struct.on('invalid', function () {
    if (this.validity.patternMismatch) {
      if (struct.val().match(/[\[\]{}]/)) {
        this.setCustomValidity('Pseudoknots are not supported');
      } else this.setCustomValidity('Structure must be composed of the characters "(", ")", and "."');
    }
  });

  // If the page has been reloaded, it may have kept old (invalid) values without retriggering custom validation
  seq.trigger('input');
  struct.trigger('input');

  $('#fill-example').click(function () {
    seq.val('NNNNAAAANNNN');
    struct.val('((((....))))');
    $('[name=nsolutions]').val(5);
  });
</script>
{% endblock bodyend %}
